web_src/js/features/common-global.js

```diff
@@ -343,3 +344,27 @@ export function initGlobalButtons() {
     });
   });
 }
+
+/**
+ * Too many users set their ROOT_URL to wrong value, and it causes a lot of problems:
+ *   * Cross-origin API request without correct cookie
+ *   * Incorrect href in <a>
+ *   * ...
+ * So we check whether current URL starts with AppUrl(ROOT_URL).
+ * If they don't match, show a warning to users.
+ */
+export function checkAppUrl() {
+  const curUrl = window.location.href;
+  if (curUrl.startsWith(appUrl)) {
+    return;
+  }
+  const $pageContent = $('.page-content');
+  if (!$pageContent.length) {
+    return;
+  }
+  const $tip = $(`<div class="ui container negative message center aligned">
```

Comment:
Should these be displayed to all users?

Reply:
Why not?

Reply:
I don't think it's appropriate to annoy the user with this error when they can't fix it anyway.

Can this JS can refactored away and only use go, by using `ctx.Flash`

Reply:
It doesn't annoy any user, if a user sees this notice, the web doesn't work correctly for them. It's good for them to ask admin for help. I do not have other ideas about further changes.

Reply:
If it's being displayed to all users, then the message should be changed according to it, currently it sounds like the user has done something wrong, also then it should be i18n then IMO.

Reply:
Why a normal user can see this warning on a correct Gitea instance?

This message should only be shown for the first installation, Gitea admin must be the first one to read it. If the Gitea admin doesn't handle it, it is the admin's problem.